### ISSUE TITLE
Validate values placed in `rank=` key

### ISSUE THREAD

**Issue by cooljeanius** [wesnoth/wesnoth#7224]
### Describe the desired feature

Currently, in the [CampaignWML](https://wiki.wesnoth.org/CampaignWML) page on the wiki, the documentation of the `rank=` key says:

> **rank**: a number that determines the order of campaigns in the campaign selection menu. Lower rank campaigns appear earlier, with unranked campaigns at the end. Currently the mainline campaigns use multiples of 10 from 0 to 399, with 0-99 for Novice campaigns, 100-199 for Intermediate campaigns, and 200-399 for Expert campaigns; if you specify this, it should not be less than 400. (Note: This replaces an older convention that topped out at 50.) ***([Version 1.14.6 and later only](https://wiki.wesnoth.org/DevFeature))*** a number that determines the order of campaigns in the campaign selection menu. Lower rank campaigns appear earlier, with unranked campaigns at the end. Currently the mainline campaigns use multiples of 5 from 0 to 249, with 0-49 for Rookie campaigns, 50-99 for Novice campaigns, 100-149 for Intermediate campaigns, 150-199 for Hard campaigns, and 200-249 for Expert campaigns; if you specify this, it should not be less than 300.

However, many UMC authors ignore this guideline, causing UMC campaigns to get mixed in with the mainline campaigns at places that often seem inappropriate. As a player who installs a lot of UMC campaigns, but also would prefer to keep them separate from mainline, I would appreciate it if Wesnoth would start enforcing a minimum value. I'm not quite sure whether it would be better to do this as a warning emitted by `wmllint`, or as something that the add-on server rejects uploads over (or possibly both), however...

**Comment by ProditorMagnus** [comment]
Once you start playing addons you dont need mainline campaigns anymore, so they could just be hidden from list. Or at least, all addon campaigns should be more visible.

wmllint is unable to do this. Addon server would take performance hit and much extra complexity to do this. And even then it probably wouldnt be able to do this for all cases.

**Comment by cooljeanius** [comment]
I don't see how it would be that complex; the idea is basically just "if rank is lower than 300, then warn/reject, else do nothing"

**Comment by ProditorMagnus** [comment]
In order to calculate value of rank you need to preprocess addon. And #ifhave can change the rank after upload.

**Comment by Elvish-Hunter** [comment]
> I'm not quite sure whether it would be better to do this as a warning emitted by wmllint, or as something that the add-on server rejects uploads over (or possibly both), however...

I'd say wmllint. We already had three different policies, and it's easier to update that script than the engine.

> I don't see how it would be that complex; the idea is basically just "if rank is lower than 300, then warn/reject, else do nothing"

Indeed it isn't complex at all, and I already have an idea about how to implement it.

> In order to calculate value of rank you need to preprocess addon. And #ifhave can change the rank after upload.

Well, they say that *perfect is the enemy of good*. It's true that add-ons can do all sorts of crazy things with the preprocessor, but most of them don't; and for those that do, there will always be `wmllint: #ignore`.

**Comment by cooljeanius** [comment]
> > I'm not quite sure whether it would be better to do this as a warning emitted by wmllint, or as something that the add-on server rejects uploads over (or possibly both), however...
> 
> I'd say wmllint. We already had three different policies, and it's easier to update that script than the engine.

OK, adding the "WML Tools" label, then

**Comment by CelticMinstrel** [comment]
This whole argument doesn't really make sense to me, honestly. If rank is supposed to reflect the campaign's overall difficulty, then it should reflect the campaign's overall difficulty regardless of whether it's a mainline campaign. On the other hand, if rank is meant as an arbitrary grouping mechanism, then the guideline of using only ranks over 400 makes a bit more sense, as that groups mainline campaigns separately from others. But a numeric rank isn't exactly a great way to achieve such an outcome when you consider than any arbitrary person can give any arbitrary campaign whatever rank they want.

Perhaps it would be better to delete the rank entirely and add some new methods that better serve the actual intended purpose.

**Comment by doofus-01** [comment]
> On the other hand, if rank is meant as an arbitrary grouping mechanism, then the guideline of using only ranks over 400 makes a bit more sense, as that groups mainline campaigns separately from others. But a numeric rank isn't exactly a great way to achieve such an outcome when you consider than any arbitrary person can give any arbitrary campaign whatever rank they want.

Yep, that's the issue.  Ranking everything by difficulty would be nice, but that's practically impossible, who would even try?

> I'm not quite sure whether it would be better to do this as a warning emitted by wmllint, or as something that the add-on server rejects uploads over (or possibly both), however...

Updating wmllint doesn't seem like much help, how many authors have just the right amount of ignorance to use misuse rank and yet use wmllint at all?  The authors who both misuse rank and use wmllint are probably doing so intentionally.

**Comment by CelticMinstrel** [comment]
> Ranking everything by difficulty would be nice, but that's practically impossible, who would even try?

Well, it sounds like that's the intent of the use of rank within the mainline campaigns, at least. We could attempt to standardize a difficulty scale and ask addon maintainers to do their best to adhere to it. Of course, it's true that there's no guaranteed that it's accurate, and I don't know if it's worth the attempt.

If it's arbitrary grouping we want, we could have something like a `group_key`, which would be a unique key that causes campaigns to be grouped. In essence, campaigns with a matching key would be shown out of the natural sort order, grouped beneath the first one with the same key. Or, there could be a `sort_key` which is used in place of the campaign's name if it exist, meaning they're grouped only when sorting by name. Or, there could be a `group_name` which is a translatable name for the group, and the group is sorted as a special "unit" which expands to contain the campaigns within it.

If using that approach, mainline could use it to keep mainline campaigns ordered at the top. Or, we could add an `is_mainline=yes` key which forces the campaign to the top, and reject any add-ons that contain that key (maybe even if setting it to false?).

**Comment by cooljeanius** [comment]
One other solution here I guess would be to just update the documentation on the wiki

**Comment by Elvish-Hunter** [comment]
In c1a074e (not backported to 1.16 yet) I updated wmllint to validate rank values. I didn't automatically close this issue in case someone wants to implement another solution.

**Comment by cooljeanius** [comment]
> In [c1a074e](https://github.com/wesnoth/wesnoth/commit/c1a074ea41a70e5044b5f81dfce654e436db7e8c) (not backported to 1.16 yet) I updated wmllint to validate rank values. I didn't automatically close this issue in case someone wants to implement another solution.

Thanks!

**Comment by Wedge009** [comment]
Just noting that with c1a074ea41a70e5044b5f81dfce654e436db7e8c, `wmllint`  now complains about the tutorial because it is assigned rank=1, which obviously isn't a multiple of 5.

**Comment by Elvish-Hunter** [comment]
`wmllint` is actually right in these three cases, so what do we want to do now? Do we prefer to change the ranking in the affected campaigns (for example, by giving `rank=0` to the tutorial), or should the rule be changed (and `wmllint` updated accordingly)?

**Comment by CelticMinstrel** [comment]
I can't think of any reason why the tutorial couldn't just have a rank of 0? And EI and WoF should probably be updated too.

**Comment by Wedge009** [comment]
In order to maintain the multiple-of-5 rule, the campaigns 'around' EI and WoF will need to be updated too.

**Comment by stevecotton** [comment]
Does updating multiple campaigns to preserve the multiple-of-5 rule make sense? Given that only mainline campaigns should have numbers in that range, using multiples of 5 seems like a way to avoid having to change multiple campaigns at once, because it leaves space to rearrange a mainline campaign between two other campaigns. But, if we start saying "it must be a multiple of 5, even if that means changing other campaigns" then we lose the advantage of using multiples.

**Comment by Wedge009** [comment]
That echoes my thinking. Like a library system to leave room for additional numbers for subsequent entries - but because the new rules follow the wiki remarks to the letter, we seem to be enforcing a multiple-of-5 without apparent reason.

**Comment by cooljeanius** [comment]
> That echoes my thinking. Like a library system to leave room for additional numbers for subsequent entries - but because the new rules follow the wiki remarks to the letter, we seem to be enforcing a multiple-of-5 without apparent reason.

yeah the wiki remarks can always be updated if we decide it's the rules we want to change instead

**Comment by cooljeanius** [comment]
So, I think this is done, isn't it? It caught a bad `rank=` key in Dawn of Thunder when I was last updating it: https://github.com/cooljeanius/dawn_of_thunder/commit/c6f47b57feb80eee2a238481ed4bd633be16dd6e 
OK to close?

**Comment by stevecotton** [comment]
It's now a discussion about whether that was the correct solution.

**Comment by stevecotton** [comment]
My feeling is still to treat it as a bug - I guess we could do an upvotes/downvotes for that.

**Comment by Wedge009** [comment]
Having spurious warnings added to the output of wmllint is not great, in my opinion - there's too many ignored warnings as it already stands. I appreciate the intention of the ranking validation, but the multiple-of-5 rule doesn't seem that useful if it's just going to cause more work.

**Comment by cooljeanius** [comment]
On second thought, this might not actually be resolved to my satisfaction after all; see czyzby/wesnoth-wml-linter#8.
I think we might want to revisit this proposal from @CelticMinstrel to add an `is_mainline` key:

> Or, we could add an `is_mainline=yes` key which forces the campaign to the top, and reject any add-ons that contain that key

**Comment by cooljeanius** [comment]
10733b1 by @Elvish-Hunter might have re-fixed this; I'll have to test...

**Comment by cooljeanius** [comment]
Although, I guess if #8321 goes through, then this would no longer be necessary...

**Comment by cooljeanius** [comment]
So, I see that #8455 was supposed to have closed this, but I think it'd be worth keeping this open until #8321 is merged.
